src/lib.rs

```diff
@@ -91,25 +99,77 @@ fn get_output_mode(args: &Args) -> OutputMode {
+
+impl MaybeInitializedCaptureIndex {
+    fn mark_failed(&self) {
+        self.0.store(u32::MAX - 1, Ordering::Relaxed);
+    }
+
+    pub fn get(&self) -> Result<Option<u32>, ()> {
+        let loaded = self.0.load(Ordering::Relaxed);
+        match loaded {
+            loaded if loaded == u32::MAX => Ok(None),
+            loaded if loaded == u32::MAX - 1 => Err(()),
+            loaded => Ok(Some(loaded)),
+        }
+    }
+
+    pub fn get_or_initialize(&self, query: &Query, capture_name: Option<&str>) -> Result<u32, ()> {
+        if let Some(already_initialized) = self.get()? {
+            return Ok(already_initialized);
+        }
+        let capture_index = match capture_name {
+            None => 0,
+            Some(capture_name) => {
+                let capture_index = query.capture_index_for_name(capture_name);
+                if capture_index.is_none() {
+                    self.mark_failed();
+                    fail(&format!("invalid capture name '{}'", capture_name));
+                }
+                capture_index.unwrap()
+            }
+        };
+        self.set(capture_index);
+        Ok(capture_index)
+    }
+
+    fn set(&self, capture_index: u32) {
+        self.0.store(capture_index, Ordering::Relaxed);
+    }
+}
+
+impl Default for MaybeInitializedCaptureIndex {
+    fn default() -> Self {
+        Self(AtomicU32::new(u32::MAX))
+    }
+}
+
 pub fn run(args: Args) {
     let query_source = match args.query_args.path_to_query_file.as_ref() {
         Some(path_to_query_file) => fs::read_to_string(path_to_query_file).unwrap(),
         None => args.query_args.query_source.clone().unwrap(),
     };
-    let supported_language = args.language.get_language();
-    let language = supported_language.language();
-    let query = Arc::new(get_query(&query_source, language));
-    let capture_index = args.capture_name.as_ref().map_or(0, |capture_name| {
-        query
-            .capture_index_for_name(capture_name)
-            .expect(&format!("Unknown capture name: `{}`", capture_name))
-    });
+    let specified_supported_language = args.language.map(|language| language.get_language());
+    let query_or_failure_by_language: Mutex<HashMap<SupportedLanguageName, Option<Arc<Query>>>> =
+        Default::default();
+    let capture_index = MaybeInitializedCaptureIndex::default();
     let output_mode = get_output_mode(&args);
     let buffer_writer = BufferWriter::stdout(ColorChoice::Never);
 
-    get_project_file_walker(&*supported_language, &args.use_paths())
+    get_project_file_walker(specified_supported_language.as_deref(), &args.use_paths())
         .into_parallel_iterator()
         .for_each(|project_file_dir_entry| {
+            let language = maybe_supported_language_from_path(project_file_dir_entry.path())
+                .expect("Walker should've been pre-filtered to just supported file types");
+            let query = return_if_none!(get_and_cache_query_for_language(
```

Comment:
Reading this I like this macro better than the boilerplate of
```
let whatever = ...;
if whatever.is_none() {
    return;
}
let whatever = whatever.unwrap();
```
or:
```
let whatever = match ... {
    None => return,
    Some(whatever) => whatever,
};
```
but it would be nice in terms of "noisiness"/"precedence" to write it postfix eg:
```
let query = ...
    .return_if_none!();
```
Looks like that is a [proposal](https://github.com/rust-lang/rfcs/pull/2442)

Or I guess if this function returned an `Option`/`Result` then could use `?`?